### PR TITLE
Pubsub

### LINKED ISSUE #122: Migrate webhookHandler to functions v2
https://firebase.google.com/docs/functions/2nd-gen-upgrade

### LINKED ISSUE #152: [Bug] Image messages being received more than once
Occasionally, but seemingly not always, image messages are being received multiple times by the system.

Meaning multiple instances and sometimes messages get created, all from just 1 image message being sent in.

Users then get back multiple replies to the same message.

Managed to replicate in dev but does not always happen.

Still unsure why this is the case. Current guess is due to time taken for OCR - perhaps WhatsApp webhook fires retries instead of waiting ~15seconds ++ (expected time for image instance handling) for a HTTP 200? Need to look into this further. 

WhatsApp Retries documentation: https://developers.facebook.com/docs/whatsapp/cloud-api/guides/set-up-webhooks

> Webhook Delivery Failure
If we send a webhook request to your endpoint and your server responds with an HTTP status code other than 200, or if we are unable to deliver the webhook for another reason, we will keep trying with decreasing frequency until the request succeeds, for up to 7 days. Note that retries will be sent to all apps that have subscribed to webhooks (and their appropriate fields) for the WhatsApp Business Account. This can result in duplicate webhook notifications.

### PR DESCRIPTION
**Pull request by sarge1989**
### Changes

- Created a webhookHandlerV2
  - Uses [cloud functions v2](https://firebase.google.com/docs/functions/2nd-gen-upgrade)
  - webhookHandlerV2 only publishes messages to 2 topics now
    - Events from checker WhatsApp number result in messages published to "checkerEvents" topic
      - This topic can eventually be removed once we have the Telegram webapp which can CRUD the firestore directly
    - Events from user WhatsApp number result in messages published to "userEvents" topic 
- Added pubsub and pubsub emulators to project
- Created onUserPublish and onCheckerPublish handlers to replace the current webhookHandler's processing
  - Uses [cloud functions v2](https://firebase.google.com/docs/functions/2nd-gen-upgrade)
  - Functionally same as previously
- Turned OCR back on

### Issues Resolved
Resolves #122  
Resolves #152 

**Important Note:**
After prod deployment, go to https://console.cloud.google.com/cloudpubsub/subscription/list?project=checkmate-373101 and
- Change acknowledgement deadline to 120 for user and 60 for checker
- Activate Exponential Backoff
- Activate dead lettering and set max delivery attempts to 75, which will mean retrying for around 12 hrs

These cant be set within the function signature (see [this](https://dev.to/danielsc/firebase-function-retries-with-pubsub-3jf9))

### CODE REVIEW

**Comment by sarge1989** [comment]
@stonefruit can help me set up the pubsub emulators in the integration test environment?

Also note the endpoint now changes to /webhookHandlerV2. I have edited the postman_collection.json already - am assuming this is the only place where it needs to change.

File: integration-tests/package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "test": "newman run checkmate.postman_collection.json -g ./env.json",
+    "test": "newman run checkmate.postman_collection.json -g ./env.json --delay-request 2000",
```
Comment (stonefruit) [inline integration-tests/package.json:7]: This seems to affect every request, however only some requests require delay. The test suite with this runs in 12 mins vs 7 mins.
Did you happen to add many new tests (`integration-tests/checkmate.postman_collection.json` diff too large to view) that might increase this timing? Or is it perhaps from the delays?

File: functions/src/definitions/webhookHandler.ts [hunk 2]
```diff
@@ -70,16 +61,19 @@ app.post("/whatsapp", async (req, res) => {
             (type == "button" || type == "interactive" || type == "text") &&
             phoneNumberId === checkerPhoneNumberId
           ) {
-            //when live, can check against WABA id instead
-            await checkerHandlerWhatsapp(message)
+            //put into checker queue
```
Comment (stonefruit) [inline functions/src/definitions/webhookHandler.ts:64]: Not a queue. pubsub messages are consumed once the subber touches it, and not when a success is returned.

File: functions/src/definitions/webhookHandler.ts [hunk 3]
```diff
@@ -50,14 +49,6 @@ app.post("/whatsapp", async (req, res) => {
         phoneNumberId === checkerPhoneNumberId ||
         phoneNumberId === userPhoneNumberId
       ) {
-        if (
```
Comment (stonefruit) [inline functions/src/definitions/webhookHandler.ts:53]: nice. lol

**Review by stonefruit — APPROVED** [review]
Just some comments to note for future.

File: integration-tests/package.json (hunk 1 above)
Comment (sarge1989) [reply integration-tests/package.json:7]: it's from the delays! haha I was lazy to individually add delays to those that are async. Is there a better way?

Comment (stonefruit) [reply integration-tests/package.json:7]: Nah. Just leave it for now. We can figure out how to optimise when it reaches 1hr. Lol

File: functions/src/definitions/webhookHandler.ts (hunk 2 above)
Comment (sarge1989) [reply functions/src/definitions/webhookHandler.ts:64]: Isit? I think it consumes on success only leh, see here:

https://cloud.google.com/pubsub/docs/push

> Here's a brief description of the workflow that references Figure 3:
> 
> The Pub/Sub server sends each message as an HTTPS request to the subscriber client at a pre-configured endpoint. This request is shown as a PushRequest in the image.
> The endpoint acknowledges the message by returning an HTTP success status code. A non-success response indicates that Pub/Sub must resend the messages. This response is shown as a PushResponse in the image.
> Pub/Sub dynamically adjusts the rate of push requests based on the rate at which it receives success responses.

**Pull request merged**